scripts/openstack/launch.py

```diff
@@ -0,0 +1,119 @@
+    num_envs = len(config["ENVS"])
+    assert num_machines >= num_seeds * num_approaches * num_envs
+    # Loop over seeds.
+    available_machines = list(machines)
+    start_seed = config["START_SEED"]
+    cmd_args = config["ARGS"]
+    for seed in range(start_seed, start_seed + num_seeds):
+        # Loop over approaches.
+        for approach_exp_id, approach_config in config["APPROACHES"].items():
+            approach = approach_config["NAME"]
+            # Loop over envs.
+            for env_exp_id, env_config in config["ENVS"].items():
+                env = env_config["NAME"]
+                # Create the experiment ID.
+                experiment_id = f"{env_exp_id}-{approach_exp_id}"
+                experiment_flags = config["FLAGS"].copy()
+                experiment_flags.update(approach_config["FLAGS"])
+                experiment_flags.update(env_config["FLAGS"])
+                # Launch the experiment for this approach / env / seed.
+                logfile = _create_logfile(experiment_id, approach, env, seed)
+                cmd = _create_cmd(experiment_id, approach, env, seed, cmd_args,
+                                  experiment_flags)
+                # One experiment per machine.
+                machine = available_machines.pop()
+                _launch_experiment(cmd, machine, logfile, args.sshkey)
+
+
+def _create_logfile(experiment_id: str, approach: str, env: str,
+                    seed: int) -> str:
+    return f"logs/{env}__{approach}__{experiment_id}__{seed}.log"
+
+
+def _create_cmd(experiment_id: str, approach: str, env: str, seed: int,
+                args: Sequence[str], flags: Dict) -> str:
+    arg_str = " ".join(f"--{a}" for a in args)
+    flag_str = " ".join(f"--{f} {v}" for f, v in flags.items())
+    cmd = f"python3.8 src/main.py --env {env} --approach {approach} " + \
+              f"--seed {seed} --experiment_id {experiment_id} {arg_str} " + \
+              f"{flag_str}"
+    return cmd
+
+
+def _run_cmds_on_machine(cmds: Sequence[str], machine: str,
+                         ssh_key: str) -> None:
+    host = f"ubuntu@{machine}"
+    ssh_cmd = f"ssh -tt -i {ssh_key} -o StrictHostKeyChecking=no {host}"
+    server_cmd_str = "\n".join(cmds + ["exit"])
+    final_cmd = f"{ssh_cmd} << EOF\n{server_cmd_str}\nEOF"
+    response = subprocess.run(final_cmd,
+                              stdout=subprocess.DEVNULL,
+                              stderr=subprocess.STDOUT,
+                              shell=True,
+                              check=False)
+    if response.returncode != 0:
+        raise RuntimeError(f"Command failed: {final_cmd}")
+
+
+def _launch_experiment(cmd: str, machine: str, logfile: str,
+                       ssh_key: str) -> None:
+    print(f"Launching on machine {machine}: {cmd}")
+    server_cmds = [
+        # Prepare the predicators directory.
+        "cd ~/predicators",
+        "mkdir -p logs",
+        "git checkout master",
```

Comment:
is it awkward that you can only run master? this could be part of the config instead right?